src/canonicalize_and_process.rs

```diff
@@ -35,40 +37,52 @@ pub fn canonicalize_hostname(value: &str) -> Result<String, ParseError> {
 }
 
 // Ref: https://wicg.github.io/urlpattern/#canonicalize-a-port
-pub fn canonicalize_port(_value: &str) -> Result<String, ParseError> {
-  todo!()
-}
-
-// Ref: https://wicg.github.io/urlpattern/#canonicalize-a-port
-pub fn canonicalize_port_with_protocol(
-  _value: &str,
-  _protocol: &str,
+pub fn canonicalize_port(
+  value: &str,
+  protocol: Option<&str>,
 ) -> Result<String, ParseError> {
-  todo!()
+  if let Some(protocol) = protocol {
+    let mut url =
+      url::Url::parse(&format!("{}://dummy.test", protocol)).unwrap(); // TODO: dont unwrap, instead ParseError
+    url::quirks::set_port(&mut url, value).unwrap(); // TODO: dont unwrap, instead ParseError
+    Ok(url::quirks::port(&url).to_string())
+  } else {
+    // If no protocol is given, the url can not have a username/password/port, so
+    // we can always return the empty string
+    Ok("".to_string())
```

Comment:
I think you should just use `String::new()`, no need to create an `&str` then convert it.